File: api-java-sql/src/main/java/com/nebulahotel/controller/ReservasController.java
package com.nebulahotel.controller;

import com.nebulahotel.model.Reservas;
import com.nebulahotel.repository.ReservasRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

@CrossOrigin
@RestController
@RequestMapping("/reservas")
public class ReservasController {

    private final ReservasRepository reservasRepository;

    public ReservasController(ReservasRepository reservasRepository) {
        this.reservasRepository = reservasRepository;
    }

    @GetMapping
    public List<Reservas> listarTodas() {
        return reservasRepository.findAll();
    }

    @GetMapping("/{id}")
    public ResponseEntity<Reservas> buscarPorId(@PathVariable Integer id) {
        return reservasRepository.findById(id)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @GetMapping("/ativas")
    public List<Reservas> listarAtivas() {
        return reservasRepository.findReservasAtivas();
    }

    @PostMapping
    public ResponseEntity<?> criar(@RequestBody Reservas reserva) {
        if (reserva.getIdReserva() == null) {
            return ResponseEntity.badRequest().body(Map.of("erro", "O campo idReserva é obrigatório."));
        }

        if (reservasRepository.existsById(reserva.getIdReserva())) {
            return ResponseEntity.status(HttpStatus.CONFLICT)
                    .body(Map.of("erro", "Já existe reserva com este idReserva."));
        }

        if (reserva.getDataReserva() == null) {
            reserva.setDataReserva(LocalDateTime.now());
        }

        Reservas salvo = reservasRepository.save(reserva);
        return ResponseEntity.status(HttpStatus.CREATED).body(salvo);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Reservas> atualizar(@PathVariable Integer id, @RequestBody Reservas reservaAtualizada) {
        if (!reservasRepository.existsById(id)) {
            return ResponseEntity.notFound().build();
        }

        reservaAtualizada.setIdReserva(id);
        Reservas salvo = reservasRepository.save(reservaAtualizada);
        return ResponseEntity.ok(salvo);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> remover(@PathVariable Integer id) {
        if (!reservasRepository.existsById(id)) {
            return ResponseEntity.notFound().build();
        }

        reservasRepository.deleteById(id);
        return ResponseEntity.noContent().build();
    }
}

File: api-kotlin-nosql/src/main/kotlin/com/nebulahotel/controller/ReservasController.kt
package com.nebulahotel.controller

import com.nebulahotel.model.Reservas
import com.nebulahotel.repository.ReservasRepository
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.time.LocalDateTime

@RestController
@RequestMapping("/reservas")
class ReservasController(
    private val reservasRepository: ReservasRepository
) {

    @GetMapping
    fun listarTodas(): List<Reservas> = reservasRepository.findAll()

    @GetMapping("/{id}")
    fun buscarPorId(@PathVariable id: Int): ResponseEntity<Reservas> {
        val reserva = reservasRepository.findById(id)
        return if (reserva.isPresent) ResponseEntity.ok(reserva.get()) else ResponseEntity.notFound().build<Reservas>()
    }

    @GetMapping("/ativas")
    fun listarAtivas(): List<Reservas> = reservasRepository.findByStatusIn(listOf("confirmada", "ativa", "em_andamento"))

    @PostMapping
    fun criar(@RequestBody reserva: Reservas): ResponseEntity<Any> {
        if (reservasRepository.existsById(reserva.idReserva)) {
            return ResponseEntity.status(HttpStatus.CONFLICT)
                .body(mapOf("erro" to "Já existe reserva com este idReserva."))
        }

        val salvo = reservasRepository.save(
            reserva.copy(dataReserva = reserva.dataReserva ?: LocalDateTime.now())
        )
        return ResponseEntity.status(HttpStatus.CREATED).body(salvo)
    }

    @PutMapping("/{id}")
    fun atualizar(@PathVariable id: Int, @RequestBody reservaAtualizada: Reservas): ResponseEntity<Reservas> {
        if (!reservasRepository.existsById(id)) {
            return ResponseEntity.notFound().build<Reservas>()
        }
        val salvo = reservasRepository.save(reservaAtualizada.copy(idReserva = id))
        return ResponseEntity.ok(salvo)
    }

    @DeleteMapping("/{id}")
    fun remover(@PathVariable id: Int): ResponseEntity<Void> {
        if (!reservasRepository.existsById(id)) {
            return ResponseEntity.notFound().build()
        }
        reservasRepository.deleteById(id)
        return ResponseEntity.noContent().build()
    }
}
